---
title: "FLAMES"
author: "Oliver Voogd"
package: FLAMES                
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{FLAMES}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## FLAMES Vignette
This vignette will detail the process of running the FLAMES pipeline. It uses the bulk data pipeline (`bulk_long_pipeline`) as an example, however this vignette will be useful for those running the other FLAMES entry point, the single cell pipeline (`sc_long_pipeline`).

### Installation
To install FLAMES, run the following:
```{r eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("FLAMES")
```

### Environment setup
To get started, the pipeline needs access to a gene annotation file in gff3 or gtf format, a directory containing one or more fastq files (which will be merged as pre-processing), a genome fasta file, as well as the file path to minimap2, and the file path to the directory to hold output files.

For this vignette, these files are downloaded from GitHub using BiocFileCache.
```{r eval=TRUE, echo=TRUE}
# download required files using BiocFileCache
temp_path <- tempfile()
bfc <- BiocFileCache::BiocFileCache(temp_path, ask=FALSE)
file_url <- 
  "https://raw.githubusercontent.com/OliverVoogd/FLAMESData/master/data"
annot <- bfc[[names(BiocFileCache::bfcadd(bfc, "Annotation", 
                                          paste(file_url, 
                                                "SIRV_isoforms_multi-fasta-annotation_C_170612a.gtf", 
                                                sep="/")))]] # [[ notation is used to get the local file path of the downloaded file
genome_fa <- bfc[[names(BiocFileCache::bfcadd(bfc, 
                                              "Genomefa", 
                                              paste(file_url, 
                                                    "SIRV_isoforms_multi-fasta_170612a.fasta", 
                                                    sep="/")))]]

# download the two fastq files, move them to a folder to be merged together
fastq1 <- bfc[[names(BiocFileCache::bfcadd(bfc, "Fastq1", paste(file_url, "fastq/sample1.fastq.gz", sep="/")))]]
fastq2 <- bfc[[names(BiocFileCache::bfcadd(bfc, "Fastq2", paste(file_url, "fastq/sample2.fastq.gz", sep="/")))]]
fastq_dir <- paste(temp_path, "fastq_dir", sep="/") # the downloaded fastq files need to be in a directory to be merged together
dir.create(fastq_dir)
file.copy(c(fastq1, fastq2), fastq_dir)
unlink(c(fastq1, fastq2)) # the original files can be deleted

# setup other environment variables
config_file <- system.file("extdata/SIRV_config_default.json", package="FLAMES") # the configuration file is included with the FLAMES package
outdir <- tempfile() # create a temporary output directory
if (!dir.exists(outdir)) {
    dir.create(outdir)
}
```

`config_file` is an optional argument which can be given to both `bulk_long_pipeline` and `sc_long_pipeline` in order to customise the execution of the pipelines. It is expected as a JSON file, and an example can be found at `r system.file("extdata/SIRV_config_default.json", package="FLAMES")`. 
If `config_file` is not given, the pipeline can instead be customised by altering any of the optional arguments the pipeline allows. These customisations are then stored in a JSON file saved in the specified out directory, which allows for easier reproduction of pipeline execution. More information on the optional arguments at the contents of the JSON configuration file can be found by running `?create_config` and `?bulk_long_pipeline`.
This vignette uses the default configuration file.

### FLAMES execution
Once the environment has been setup, the pipeline can be executed by running:
```{r eval=FALSE}
library(FLAMES)
summarizedExperiment <- bulk_long_pipeline(annot=annot, fastq=fastq_dir, outdir=temp_path, 
                                           genome_fa=genome_fa, minimap2_dir=mm2_dir,
                                           config_file = config_file)
```

### FLAMES termination
The directory `outdir` now contains several output files returned from this pipeline. The output files generated by this pipeline are:

\itemize{
  \item{transcript_count.csv.gz}{ - a transcript count matrix (also contained in the SummarizedExperiment)}
  \item{isoform_annotated.filtered.gff3}{ - isoforms in gff3 format (also contained in the SummarizedExperiment)}
  \item{transcript_assembly.fa}{ - transcript sequence from the isoforms}
  \item{align2genome.bam}{ - sorted BAM file with reads aligned to genome}
  \item{realign2transcript.bam}{ - sorted realigned BAM file using the transcript_assembly.fa as reference}
  \item{tss_tes.bedgraph}{ - TSS TES enrichment for all reads (for QC)}
 }

The pipeline also returns a SummarizedExperiment  or SingleCellExperiment object, depending on the pipeline run, containing the data from `transcript_count.csv.gz`and `isoform_annotated.filtered.gff3`.


### FLAMES on Windows
Due to FLAMES requiring minimap2 to complete the pipeline, the straight FLAMES pipeline functions `bulk_long_pipeline` and `sc_long_pipeline` won't run on a Windows OS. To overcome this issue, the vignette 'Vignette for FLAMES on Windows' describes the alternate method of running the FLAMES pipelines, which requires acccess to minimap2 on an external system

#### Session Info
``` {r echo=FALSE}
utils::sessionInfo()
```