#' Pipeline for Multi-sample Single Cell Data
#'
#' @md
#'
#' @description
#' Semi-supervised isoform detection and annotation for long read data.
#' This variant is for multi-sample single cell data. By default, this pipeline demultiplexes input
#' fastq data (\code{match_cell_barcode = TRUE}). Specific parameters relating to
#' analysis can be changed either through function arguments, or through a
#' configuration JSON file.
#'
#' @details
#' By default FLAMES use minimap2 for read alignment. After the genome alignment step (\code{do_genome_align}), FLAMES summarizes the alignment for each read in every sample by grouping reads
#' with similar splice junctions to get a raw isoform annotation (\code{do_isoform_id}). The raw isoform
#' annotation is compared against the reference annotation to correct potential splice site
#' and transcript start/end errors. Transcripts that have similar splice junctions
#' and transcript start/end to the reference transcript are merged with the
#' reference. This process will also collapse isoforms that are likely to be truncated
#' transcripts. If \code{isoform_id_bambu} is set to \code{TRUE}, \code{bambu::bambu} will be used to generate the updated annotations (Not implemented for multi-sample yet).
#' Next is the read realignment step (\code{do_read_realign}), where the sequence of each transcript from the update annotation is extracted, and
#' the reads are realigned to this updated \code{transcript_assembly.fa} by minimap2. The
#' transcripts with only a few full-length aligned reads are discarded (Not implemented for multi-sample yet).
#' The reads are assigned to transcripts based on both alignment score, fractions of
#' reads aligned and transcript coverage. Reads that cannot be uniquely assigned to
#' transcripts or have low transcript coverage are discarded. The UMI transcript
#' count matrix is generated by collapsing the reads with the same UMI in a similar
#' way to what is done for short-read scRNA-seq data, but allowing for an edit distance
#' of up to 2 by default. Most of the parameters, such as the minimal distance to splice site and minimal percentage of transcript coverage
#' can be modified by the JSON configuration file (\code{config_file}).
#'
#' @param fastqs A vector containing the paths to each fastq files. If \code{in_bams} is not provided, this argument can also
#' be provided as the path to the folder containing the fastq files. Each fastq file will be treated as a sample.
#' @param in_bams Optional vector containing file paths the  bam files to use instead of fastq file (skips initial alignment step).
#' The order of the bam files need to mach the order in \code{fastqs}.
#' @inheritParams sc_long_pipeline
#' @return \code{sc_long_pipeline} returns a SingleCellExperiment object, containing a count
#' matrix as an assay, gene annotations under metadata, as well as a list of the other
#' output files generated by the pipeline. The pipeline also outputs a number of output
#' files into the given \code{outdir} directory. These output files generated by the pipeline are:
#' \itemize{
#'  \item{transcript_count.csv.gz}{ - a transcript count matrix (also contained in the SingleCellExperiment)}
#'  \item{isoform_annotated.filtered.gff3}{ - isoforms in gff3 format (also contained in the SingleCellExperiment)}
#'  \item{transcript_assembly.fa}{ - transcript sequence from the isoforms}
#'  \item{align2genome.bam}{ - sorted BAM file with reads aligned to genome}
#'  \item{realign2transcript.bam}{ - sorted realigned BAM file using the transcript_assembly.fa as reference}
#'  \item{tss_tes.bedgraph}{ - TSS TES enrichment for all reads (for QC)}
#' }
#'
#' @details The default parameters can be changed either through the function
#' arguments are through the configuration JSON file \code{config_file}. the \code{pipeline_parameters}
#' section specifies which steps are to be executed in the pipeline - by default, all
#' steps are executed. The \code{isoform_parameters} section affects isoform detection - key
#' parameters include:
#' \itemize{
#'  \item{\code{Min_sup_cnt}}{ which causes transcripts with less reads aligned than
#' it's value to be discarded}
#'  \item{\code{MAX_TS_DIST}}{ which merges transcripts with the same intron
#' chain and TSS/TES distace less than \code{MAX_TS_DIST}}
#'  \item{\code{strand_specific}}{ which specifies if reads are in the same strand as the mRNA (1),
#' or the reverse complemented (-1) or not strand specific (0), which results in
#' strand information being based on reference annotation.}
#' }
#'
#' @seealso
#' [bulk_long_pipeline()] for bulk long data,
#' [SingleCellExperiment()] for how data is outputted
#'
#' @importFrom dplyr group_by summarise_at slice_max filter
#' @importFrom magrittr "%>%"
#' @importFrom SingleCellExperiment SingleCellExperiment reducedDimNames logcounts
#' @importFrom SummarizedExperiment rowData colData rowData<- colData<- rowRanges rowRanges<-
#' @importFrom BiocGenerics cbind colnames rownames start end
#' @importFrom utils read.csv read.table file_test
#' @importFrom jsonlite fromJSON
#'
#' @example inst/examples/pipeline_example.R
#' @export
sc_long_multisample_pipeline <-
    function(annotation,
             fastqs,
             outdir,
             genome_fa,
             minimap2_dir = NULL,
             reference_csv,
             match_barcode = TRUE,
             config_file = NULL) {
        checked_args <- check_arguments(
            annotation,
            fastqs,
            genome_bam = NULL,
            outdir,
            genome_fa,
            minimap2_dir,
            config_file
        )

        config <- checked_args$config

        # check fastqs
        if (length(fastqs) == 1) {
            if (file_test("-f", fastqs)) {
                stop("Only one fastq file provided, did you meant to used the single-sample pipeline (FLAMES::sc_long_pipeline) ?")
            }

            fastqs <- file.path(fastqs, list.files(fastqs))[grepl("\\.(fq|fastq)$", list.files(fastqs))]
            cat("Fastq files found:\n")
            cat(paste0(fastqs, sep = "\n"))

            if (length(fastqs) <= 1) {
                stop(length(fastqs), " .fq or .fastq file(s) found\n")
            }

            if (match_barcode) {
                stop("If \"match_barcode\" set to TRUE, argument \"fastqs\" must be a list of fastq files, with the same order in \"reference_csv\"\nYou can also demultiplex the reads with \"FLAMES::match_cell_barcode_cpp\"")
            }
        } else if (any(!file.exists(fastqs))) {
            stop("Please make sure all fastq files exist.")
        }

        samples <- gsub("\\.(fastq|fq)(\\.gz)?$", "", basename(fastqs))

        if (match_barcode && length(reference_csv) >= 1) {
            if (!all(file.exists(reference_csv))) {
                stop("Please make sure all reference_csv file exists.\n")
            }
            if (length(reference_csv) == 1) {
                reference_csv <- rep(reference_csv, length(fastqs))
            }
            if (length(reference_csv) != length(fastqs)) {
                stop(length(reference_csv), " barcode allow-lists provided while there are ", length(fastqs), "fastq file. Please either provide one allow-list per sample, or one allow-list for all samples.")
            }
            infqs <- file.path(outdir, paste(samples, "matched_reads.fastq.gz", sep = "_"))
            bc_stats <- file.path(outdir, paste(samples, "matched_barcode_stat", sep = "_"))
            for (i in length(fastqs)) {
                match_cell_barcode_cpp(
                    fastqs[i],
                    bc_stats[i],
                    infqs[i],
                    reference_csv[i],
                    config$barcode_parameters$max_edit_distance,
                    config$barcode_parameters$UMI_length
                )
            }
        } else {
            infqs <- fastqs
        } # requesting to not match barcodes implies `fastq` has already been run through the
        # function in a previous FLAMES call

        cat("Running FLAMES pipeline...\n")

        using_bam <- FALSE
        genome_bam <- file.path(outdir, paste0(samples, "_", "align2genome.bam"))
        if (all(utils::file_test("-f", genome_bam))) {
            cat("Found all corresponding '[sample]_align2genome.bam' files, will skip initial alignment.\n")
            using_bam <- TRUE
            config$pipeline_parameters$do_genome_alignment <- FALSE
            if (!all(utils::file_test("-f", file.path(outdir, paste0(samples, "_", "align2genome.bam.bai"))))) {
                for (bam in genome_bam) {
                    Rsamtools::indexBam(bam)
                }
            }
        }


        cat("#### Input parameters:\n")
        cat(jsonlite::toJSON(config, pretty = TRUE))
        cat("gene annotation:", annotation, "\n")
        cat("genome fasta:", genome_fa, "\n")
        if (using_bam) {
            cat("input bam:", paste0(genome_bam, sep = "\n"), "\n")
        }
        cat("input fastqs:", paste0(fastqs, sep = "\n"), "\n")
        cat("output directory:", outdir, "\n")
        cat("directory containing minimap2:", minimap2_dir, "\n")

        # align reads to genome
        # if (!using_bam && config$pipeline_parameters$do_genome_alignment) {
        if (config$pipeline_parameters$do_genome_alignment) {
            cat("#### Aligning reads to genome using minimap2\n")
            for (i in 1:length(samples)) {
                cat(paste0(c("\tAligning sample ", samples[i], "...\n")))
                minimap2_align(
                    config,
                    genome_fa,
                    fastqs[i],
                    annotation,
                    outdir,
                    minimap2_dir,
                    prefix = samples[i],
                    threads = NULL
                )
            }
        } else {
            cat("#### Skip aligning reads to genome\n")
        }

        # find isofroms
        if (config$pipeline_parameters$do_isoform_identification) {
            find_isoform(annotation, genome_fa, genome_bam, outdir, config)
        }

        # realign to transcript
        if (config$pipeline_parameters$do_read_realignment) {
            cat("#### Realign to transcript using minimap2\n")
            for (i in 1:length(samples)) {
                cat(paste0(c("\tRealigning sample ", samples[i], "...\n")))
                minimap2_realign(config, fastqs[i], outdir, minimap2_dir, prefix = samples[i], threads = NULL)
            }
        } else {
            cat("#### Skip read realignment\n")
        }

        # quantification
        # TODO: implement filtering in R
        if (config$pipeline_parameters$do_transcript_quantification) {
            cat("#### Generating transcript count matrix\n")
            flames_quantify(annotation = annotation, outdir = outdir, config = config, pipeline = "sc_multi_sample")
        } else {
            cat("#### Skip transcript quantification\n")
        }

        sce_list <- as.list(1:length(samples))
        names(sce_list) <- samples
        load_genome_anno <- rtracklayer::import(annotation, feature.type = c("exon", "utr"))

        for (i in 1:length(samples)) {
            out_files <- list(
                "annotation" = annotation,
                "genome_fa" = genome_fa,
                "counts" = file.path(outdir, paste0(samples[i], "_transcript_count.csv.gz")),
                "isoform_annotated" = file.path(outdir, "isoform_annotated.gff3"),
                "transcript_assembly" = file.path(outdir, "transcript_assembly.fa"),
                "config" = config,
                "align_bam" = genome_bam[i],
                "realign2transcript" = file.path(outdir, paste0(samples[i], "_realign2transcript.bam")),
                "outdir" = outdir,
                "fsm_annotation" = file.path(outdir, "isoform_FSM_annotation.csv")
            )
            sce_list[[i]] <- generate_sc_singlecell(out_files, load_genome_anno = load_genome_anno)
        }

        return(sce_list)
    }
