#' Pipeline for Single Cell Data
#'
#' @md
#' @description NEEDED
#'
#' @inheritParams bulk_long_pipeline
#' @inherit bulk_long_pipeline details description
#' @param UMI_LEN NEEDED
#'
#' @return \code{sc_long_pipeline} returns a SingleCellExperiment object, containing a count
#' matrix as an assay, gene annotations under metadata, as well as a list of the other
#' output files generated by the pipeline. The pipeline also outputs a number of output
#' files into the given \code{outdir} directory. These output files generated by the pipeline are:
#' \itemize{
#'  \item{transcript_count.csv.gz}{ - a transcript count matrix (also contained in the SingleCellExperiment)}
#'  \item{isoform_annotated.filtered.gff3}{ - isoforms in gff3 format (also contained in the SingleCellExperiment)}
#'  \item{transcript_assembly.fa}{ - transcript sequence from the isoforms}
#'  \item{align2genome.bam}{ - sorted BAM file with reads aligned to genome}
#'  \item{realign2transcript.bam}{ - sorted realigned BAM file using the transcript_assembly.fa as reference}
#'  \item{tss_tes.bedgraph}{ - TSS TES enrichment for all reads (for QC)}
#' }
#'
#' @seealso
#' [bulk_long_pipeline()] for bulk long data,
#' [SingleCellExperiment()] for how data is outputted
#'
#' @importFrom SingleCellExperiment SingleCellExperiment
#' @importFrom utils read.csv read.table
#'
#' @export
#sc_long_pipeline <- function(annot, fastq, in_bam, outdir, genome_fa,
sc_long_pipeline <- function(annot, fastq, outdir, genome_fa,
                                minimap2_dir=NULL, downsample_ratio=1, config_file=NULL,
                                do_genome_align=TRUE, do_isoform_id=TRUE,
                                do_read_realign=TRUE, do_transcript_quanti=TRUE,
                                gen_raw_isoform=TRUE, has_UMI=FALSE, UMI_LEN=10,
                                MAX_DIST=10, MAX_TS_DIST=100, MAX_SPLICE_MATCH_DIST=10,
                                min_fl_exon_len=40, Max_site_per_splice=3, Min_sup_cnt=10,
                                Min_cnt_pct=0.01, Min_sup_pct=0.2, strand_specific=1, remove_incomp_reads=5,
                                use_junctions=TRUE, no_flank=TRUE,
                                use_annotation=TRUE, min_tr_coverage=0.75, min_read_coverage=0.75) {

    infq <- paste(outdir, "matched_reads.fastq.gz", sep="/")
    bc_stat <- paste(outdir, "matched_barcode_stat", sep="/")
    ref_csv <- "?????"
    match_cell_barcode(fastq, bc_stat, infq, ref_csv, MAX_DIST, UMI_LEN)

    #generic_long_pipeline(annot, infq, in_bam, outdir, genome_fa,
    generic_long_pipeline(annot, infq, outdir, genome_fa,
            minimap2_dir, downsample_ratio, config_file,
            do_genome_align, do_isoform_id,
            do_read_realign, do_transcript_quanti,
            gen_raw_isoform, has_UMI,
            MAX_DIST, MAX_TS_DIST, MAX_SPLICE_MATCH_DIST,
            min_fl_exon_len, Max_site_per_splice, Min_sup_cnt,
            Min_cnt_pct, Min_sup_pct, strand_specific, remove_incomp_reads,
            use_junctions, no_flank,
            use_annotation, min_tr_coverage, min_read_coverage)

    # files in the output directory:
    #   transcript_count.csv.gz   // transcript count matrix
    #   isoform_annotated.filtered.gff3 // isoforms in gff3 format
    #   transcript_assembly.fa // transcript sequence from the isoforms
    #   align2genome.bam       // sorted bam file with reads aligned to genome
    #   realign2transcript.bam // sorted realigned bam file using the
    #                            transcript_assembly.fa as reference
    #   tss_tes.bedgraph       // TSS TES enrichment for all reads (for QC)

    # this method requires testing using single cell data
    counts <- read.csv(paste0(outdir, "/transcript_count.csv.gz"))
    annot <- read.table(paste0(outdir, "/isoform_annotated.filtered.gff3"))
    colnames(annot) <- c("SequenceID", "Source", "Feature", "Start", "End", "Score", "Strand", "Phase", "Attributes")
    mdata <- list(
            "Annotations"=annot,
            "OutputFiles"=
                        list("transcript_assembly"=paste0(outdir, "/transcript_assembly.fa"),
                        "align2genome"=paste0(outdir, "/align2genome.bam"),
                        "realign2transcript"=paste0(outdir, "/realign2transcript.bam"),
                        "tss_tes"=paste0(outdir, "/tss_tes.bedgraph")
                        )
                )
    sce <- SingleCellExperiment(list("Flames Single Cell"=counts),
                                metadata=mdata)
    #return the created singlecellexperiment
    sce
}

