#' Pipeline for Single Cell Data
#'
#' @md
#' @description NEEDED
#'
#' @param annot gene annotations file in gff3  format
#' @param fastq file path to input fastq file
#' @param outdir directory to store all output files.
#' @param genome_fa genome fasta file.
#' @param minimap2_dir directory containing minimap2, k8 and paftools.js program.
#' k8 and paftools.js are used to convert gff3 to bed12.
#' @param downsample_ratio downsampling ratio if performing downsampling analysis.
#' @param config_file JSON configuration file. If specified, \code{config_file} overrides
#' all configuration parameters
#' @param do_genome_align Boolean. Specifies whether to run the genome alignment step. \code{TRUE} is recommended
#' @param do_isoform_id Boolean. Specifies whether to run the isoform identification step. \code{TRUE} is recommended
#' @param do_read_realign Boolean. Specifies whether to run the read realignment step. \code{TRUE} is recommended
#' @param do_transcript_quanti Boolean. Specifies whether to run the transcript quantification step. \code{TRUE} is recommended
#' @param gen_raw_isoform Boolean.
#' @param has_UMI Boolean. Speficies if each gene as a UMI.
#' @param MAX_DIST Numeric
#' @param MAX_TS_DIST Numeric.
#' @param MAX_SPLICE_MATCH_DIST Numeric.
#' @param min_fl_exon_len Numeric.
#' @param Max_site_per_splice Numeric.
#' @param Min_sup_cnt Numeric.
#' @param Min_cnt_pct Numeric.
#' @param Min_sup_pct Numeric.
#' @param strand_specific 1, -1 or 0. 1 indicates if reads are in the same
#' strand as mRNA, -1 indicates reads are reverse complemented, 0 indicates
#' reads are not strand specific.
#' @param remove_incomp_reads Numeric.
#' @param use_junctions Boolean.
#' @param no_flank Boolean.
#' @param use_annotation Boolean.
#' @param min_tr_coverage Numeric.
#' @param min_read_coverage Numeric.
#' @inherit bulk_long_pipeline details description
#' @param UMI_LEN NEEDED
#' @param reference_csv file path to the reference csv used for demultiplexing
#' @param match_barcode bool; specifies if demultiplexing should be performed using `FLAMES::match_cell_barcode`
#' @return \code{sc_long_pipeline} returns a SingleCellExperiment object, containing a count
#' matrix as an assay, gene annotations under metadata, as well as a list of the other
#' output files generated by the pipeline. The pipeline also outputs a number of output
#' files into the given \code{outdir} directory. These output files generated by the pipeline are:
#' \itemize{
#'  \item{transcript_count.csv.gz}{ - a transcript count matrix (also contained in the SingleCellExperiment)}
#'  \item{isoform_annotated.filtered.gff3}{ - isoforms in gff3 format (also contained in the SingleCellExperiment)}
#'  \item{transcript_assembly.fa}{ - transcript sequence from the isoforms}
#'  \item{align2genome.bam}{ - sorted BAM file with reads aligned to genome}
#'  \item{realign2transcript.bam}{ - sorted realigned BAM file using the transcript_assembly.fa as reference}
#'  \item{tss_tes.bedgraph}{ - TSS TES enrichment for all reads (for QC)}
#' }
#'
#' @seealso
#' [bulk_long_pipeline()] for bulk long data,
#' [SingleCellExperiment()] for how data is outputted
#'
#' @importFrom SingleCellExperiment SingleCellExperiment
#' @importFrom utils read.csv read.table
#'
#' @export
sc_long_pipeline <- function(annot, fastq, in_bam=NULL, outdir, genome_fa,
                                minimap2_dir=NULL, downsample_ratio=1, 
                                reference_csv, match_barcode=TRUE, config_file=NULL,
                                do_genome_align=TRUE, do_isoform_id=TRUE,
                                do_read_realign=TRUE, do_transcript_quanti=TRUE,
                                gen_raw_isoform=TRUE, has_UMI=FALSE, UMI_LEN=10,
                                MAX_DIST=10, MAX_TS_DIST=100, MAX_SPLICE_MATCH_DIST=10,
                                min_fl_exon_len=40, Max_site_per_splice=3, Min_sup_cnt=10,
                                Min_cnt_pct=0.01, Min_sup_pct=0.2, strand_specific=1, remove_incomp_reads=5,
                                use_junctions=TRUE, no_flank=TRUE,
                                use_annotation=TRUE, min_tr_coverage=0.75, min_read_coverage=0.75) {
    if (is.null(in_bam)) {
        if (match_barcode) {
                if (!file.exists(reference_csv)) stop("reference_csv must exists.")
                infq <- paste(outdir, "matched_reads.fastq.gz", sep="/")
                bc_stat <- paste(outdir, "matched_barcode_stat", sep="/")
                match_cell_barcode(fastq, bc_stat, infq, reference_csv, MAX_DIST, UMI_LEN)
        } else infq = fastq # requesting to not match barcodes implies `fastq` has already been run through the 
                            # function in a previous FLAMES call
    }

    out_files <- generic_long_pipeline(annot, infq, in_bam=in_bam, outdir, genome_fa,
            minimap2_dir, downsample_ratio, config_file,
            do_genome_align, do_isoform_id,
            do_read_realign, do_transcript_quanti,
            gen_raw_isoform, has_UMI,
            MAX_DIST, MAX_TS_DIST, MAX_SPLICE_MATCH_DIST,
            min_fl_exon_len, Max_site_per_splice, Min_sup_cnt,
            Min_cnt_pct, Min_sup_pct, strand_specific, remove_incomp_reads,
            use_junctions, no_flank,
            use_annotation, min_tr_coverage, min_read_coverage)

    sce <- generate_sc_singlecell(out_files)

    sce
}

generate_sc_singlecell <- function(out_files) {
    # this method requires testing using single cell data
    counts <- read.csv(out_files$counts)
    annot <- read.table(out_files$annot)
    colnames(annot) <- c("SequenceID", "Source", "Feature", "Start", "End", "Score", "Strand", "Phase", "Attributes")
    mdata <- list(
            "Annotations"=out_files$annot,
            "OutputFiles"=out_files
                )
    sce <- SingleCellExperiment::SingleCellExperiment(list("Flames Single Cell"=counts),
                                metadata=mdata)
    #return the created singlecellexperiment
    sce
}